File: asme/ops/serializers/comments.py
"""JSON shape for comments.

``comment`` follows the plan's shape plus ``deleted`` and ``mentions``. A
soft-deleted comment keeps its place in the thread but exposes no text.
"""

from __future__ import annotations

from asme.ops.serializers import iso, uid, user_ref


def comment(row, mentioned_users=()) -> dict:
    deleted = row.deleted_at is not None
    return {
        "id": uid(row.id),
        "entity_type": row.entity_type,
        "entity_id": uid(row.entity_id),
        "author": user_ref(row.author),
        "body": "" if deleted else row.body,
        "parent_comment_id": uid(row.parent_comment_id),
        "edited_at": iso(row.edited_at),
        "created_at": iso(row.created_at),
        "deleted": deleted,
        "mentions": [] if deleted else [user_ref(user) for user in mentioned_users],
    }

File: apps/ops-web/src/api/contracts/comments.ts
import { z } from 'zod'
import { listOf, UserRef } from './common'

/** Shape from asme/ops/serializers/comments.py. Soft-deleted comments keep their place with an empty body. */
export const Comment = z.looseObject({
  id: z.string(),
  entity_type: z.string(),
  entity_id: z.string(),
  author: UserRef.nullable(),
  body: z.string(),
  parent_comment_id: z.string().nullable().optional(),
  edited_at: z.string().nullable().optional(),
  created_at: z.string(),
  deleted: z.boolean().default(false),
  mentions: z.array(UserRef).default([]),
})
export type Comment = z.infer<typeof Comment>

export const CommentList = listOf(Comment)
export type CommentList = z.infer<typeof CommentList>

/** URL segment accepted by `/:entity/:id/comments`. */
export type CommentEntity = 'work-orders' | 'projects' | 'assets'

export const CommentInput = z.object({
  body: z.string().trim().min(1, 'Write something first.').max(8000, 'Keep comments under 8,000 characters.'),
  parent_comment_id: z.string().nullable().optional(),
})
export type CommentInput = z.infer<typeof CommentInput>
